src/modules/commander/Commander.cpp

```diff
@@ -4058,6 +4068,164 @@ void Commander::battery_status_check()
+		// addition to a sanity check using airspeed and load factor and a missing sensor data check.
+
+		// Check if sensor data is missing - assume a minimum 5Hz data rate.
+		bool data_missing = false;
+		if ((hrt_absolute_time() -_time_last_airspeed) > 200000) {
+			data_missing = true;
+		}
+
+		// Declare data stopped if not received for longer than 1 second
+		bool data_stopped = false;
+		if ((hrt_absolute_time() -_time_last_airspeed) > 1000000) {
+			data_stopped = true;
+		}
+		_time_last_airspeed = hrt_absolute_time();
+
+		// Check if the airpeed reading is lower than physically possible given the load factor
+		_airspeed_sub.update();
+		_airspeed = _airspeed_sub.get();
+		_sensor_combined_sub.update();
+		_sensor_combined = _sensor_combined_sub.get();
+		float max_lift_ratio = _airspeed.indicated_airspeed_m_s /fmaxf(_airspeed_stall.get(),1.0f);
+		max_lift_ratio *= max_lift_ratio;
+		status.load_factor_ratio = 0.95f * status.load_factor_ratio - 0.05f * (_sensor_combined.accelerometer_m_s2[2] / 9.80665f) / max_lift_ratio;
+		status.load_factor_ratio = math::constrain(status.load_factor_ratio, 0.25f, 2.0f);
+		bool load_factor_ratio_fail = status.load_factor_ratio > 1.1f;
+
+		//  Decide if the control loops should be using the airspeed data based on the length of time the
+		// airspeed data has been declared bad
+		if (_tas_check_fail || load_factor_ratio_fail || data_missing) {
+			// either load factor or EKF innovation or missing data test failure can declare the airspeed bad
+			_time_tas_bad_declared = hrt_absolute_time();
+			status.aspd_check_failing = true;
+		} else if (!_tas_check_fail && !load_factor_ratio_fail && !data_missing) {
+			// All checks must pass to declare airspeed good
+			_time_tas_good_declared = hrt_absolute_time();
+			status.aspd_check_failing = false;
+		}
+		if (!_tas_use_inhibit) {
+			// A simultaneous load factor and innovaton check fail makes it more likely that a large
+			// airspeed measurement fault has developed, so a fault should be declared immediately
+			bool both_checks_failed = _tas_check_fail && load_factor_ratio_fail;
+
+			// Because the innovation, load factor and data missing checks are subject to short duration false positives
+			// a timeout period is applied.
+			bool single_check_fail_timeout = (hrt_absolute_time() - _time_tas_good_declared) > 1000000 * (hrt_abstime)_tas_use_stop_delay.get();
+
+			if (data_stopped || both_checks_failed || single_check_fail_timeout) {
+				_tas_use_inhibit = true;
+				fault_declared = true;
+				if (data_stopped || data_missing) {
+					strcpy(_airspeed_fault_type, "MISSING");
+				} else  {
+					strcpy(_airspeed_fault_type, "FAULTY ");
+				}
+			}
+		} else if ((hrt_absolute_time() - _time_tas_bad_declared) > 1000000 * (hrt_abstime)_tas_use_start_delay.get()) {
+			_tas_use_inhibit = false;
+			fault_cleared = true;
+		}
+	}
+
+	// Do actions based on value of COM_ASPD_FS_ACT parameter
+	switch (_airspeed_fail_action.get()) {
```

Comment:
There's a fair amount of repetition here. What about combining the fault_cleared logic and status.aspd_fault_declared always set to fault_declared? Setting every flag in every case shouldn't be necessary.

Reply:
I  need to think about that. The way it is set out at the moment makes it easy for me to follow and debug. I will set the flags to false at the top and set to true by exception.